server.ts

```diff
@@ -0,0 +1,318 @@
+        return;
+      }
+
+      const decodedEnvelope = JSON.parse(item.content.toString());
+      const update: OccupancyUpdate = Object.assign(<OccupancyUpdate>{}, decodedEnvelope);
+      for (const occupancy of update.occupancy) {
+        occupancyUpdate(update.channel, occupancy);
+      }
+
+      // Remove message from queue.
+      // We do it here so that any thrown exception will re-trigger processing this message.
+      queueChannel.ack(item);
+    });
+  } catch (err) {
+    console.error('worker:', 'Queue error!', err);
+  }
+};
+
+// Fetch presence set from a list of channels.
+async function fetchPresenceSets(channels: string, updateMemberCount: boolean) {
+  const content = { "channels": channels }
+  const result = await rest.request('GET', '/presence', content);
+  for (const item of result.items) {
+    const presenceSet: PresenceSet = Object.assign(<PresenceSet>{}, item);
+    const messages = Ably.Realtime.PresenceMessage.fromEncodedArray(presenceSet.presence);
+    for (const message of messages) {
+      presenceUpdate(presenceSet.channel, message, updateMemberCount);
+    }
+  }
+}
+
+// Called when the presence member count (set using occupancy messages)
+// needs to be checked with the presence member map (filled using presence events).
+// Triggers fetching the current presence set if both are not equal.
+function checkPresenceMemberCount(channelId: string): void {
+  if (!storage.has(channelId)) {
+    return;
+  }
+  const channel = storage.get(channelId)!;
+  channel.memberCountCheckTimeout = null;
+  if (channel.expectedPresenceMembers != channel.members.size) {
+    console.log('Member count error detected for channel ' + channelId + ', occupancy: ' + channel.expectedPresenceMembers +
+      ', us: ' + channel.members.size + '; syncing');
+    fetchPresenceSets(channelId, true);
+  }
+}
+
+// A presence member key. Presence members can be uniquely identified using
+// a connection id and a client id, so this computes a hash of both.
+// This is used as a key in the storage map.
+class MemberKey {
+  hash: number;
+  constructor(clientId: string, connectionId: string) {
+    this.hash = Mmh3.murmur32Sync(clientId + connectionId);
+  }
+}
+
+// A presence member. The timestamp is used to ensure we don't override presence data
+// because we have received an earlier event.
+class Member {
+  latestTimestamp: number;
+  clientId: string;
+  connectionId: string;
+  constructor(clientId: string, connectionId: string) {
+    this.clientId = clientId;
+    this.connectionId = connectionId;
+    this.latestTimestamp = 0;
+  }
+}
+
+// Channel data.
+class Channel {
+  expectedPresenceMembers: number;
+  memberCountCheckTimeout: NodeJS.Timeout | null;
+  members: Map<number, Member>;
+  constructor() {
+    this.expectedPresenceMembers = 0;
+    this.memberCountCheckTimeout = null;
+    this.members = new Map<number, Member>();
+  }
+}
+
+// Global storage map.
```

Comment:
Can you add in a production solution this may, depending on the use case, be a database connection.

Reply:
You mean add a comment to say that in a prod env this could be a DB instead? Sure.